Shape-verifying SVG reference tests

skef opened the pull request:
There is currently very little automated verification of shape rendering for SVG elements. This pull request adds reference tests with very-close-to-equivalent path geometries and fuzzy matching. The "fuzz factors" were chosen based by putting a medium-tight margin above the worse of (Linux) Firefox and Chromium rendering. (Opera was verified to be equivalent to the latter; no other browsers were tested.)

The circles, ellipses, lines, and polygons are taken from `shapes-circle-01-t-manual.svg`, `shapes-ellipse-01-t-manual.svg`, `shapes-line-02-f-manual.svg` and `shapes-polygon-01-t-manual.svg`. `shapes-polygon-02-t-manual.svg` is also superseded. The (somewhat trivial) "strokefill" paths are taken from `render-elems-02-t-manual.svg` and `render-elems-03-t-manual.svg`. The "linejoincap" tests are taken from `paths-data-10-t-manual.svg`. The "miterlimit" test is taken from `painting-stroke-07-t-manual-ref.svg`. The "linejoin", "linecap", and "rect" tests are new but more or less supersede `painting-stroke-02-t-manual.svg`, `painting-stroke-03-t-manual.svg`, `shapes-rect-02-t-manual.svg`, and `shapes-rect-03-t-manual.svg`. (All of these existing files are in the `import` directory; none are deleted in this PR.) 

This zip archive has a record of the max channel difference and pixel counts (in `info.txt`), the rasterizations in both browsers, and a difference image for each reference test generated with ImageMagick's `compare` option. 

[images.zip](https://github.com/web-platform-tests/wpt/files/4450815/images.zip)

(This PR is a sort of rehearsal for a more interesting/future-oriented one along similar lines.)

skef committed:
Geometry-verifying fuzzy-matched SVG reference tests

skef commented:
I ran into some recent discussion on one of the mailing lists of an alternate matching heuristic using an under-placed narrow outline of similar color, along these lines:

```
<svg id="svg-root"
  width="100%" height="100%" viewBox="0 0 480 360"
  xmlns="http://www.w3.org/2000/svg"
  xmlns:xlink="http://www.w3.org/1999/xlink"
  xmlns:html="http://www.w3.org/1999/xhtml">
  <g id="testmeta">
    <title>Path With Round Caps</title>
    <html:link rel="author"
          title="Skef Iterum"
          href="https://github.com/skef/"/>
    <html:link rel="help"
          href="https://www.w3.org/TR/SVG2/painting.html#LineCaps"/>
    <html:link rel="match" href="linecap-001-ref.svg"/>
    <html:meta name="fuzzy" content="maxDifference=10;totalPixels=0-4000"/>
  </g>

  <g id="test-body-content">
    <path style="stroke:#0A0AFF;stroke-width:2;fill:none" d="M219.021 20.2002l50 5c20.1914 2.01855 35.9795 19.0801 35.9795 39.7998c0 22.0762 -17.9238 40 -40 40c-1.35742 0 -2.69824 -0.0673828 -4.02051 -0.200195l-50 -5c-20.1914 -2.01855 -35.9795 -19.0801 -35.9795 -39.7998
c0 -22.0762 17.9238 -40 40 -40c1.35742 0 2.69824 0.0673828 4.02051 0.200195z"/>
    <path fill="none" stroke="blue" stroke-width="80" stroke-linecap="round" d="M 215 60 l 50 5"/>
  </g>
</svg>

<svg id="svg-root"
  width="100%" height="100%" viewBox="0 0 480 360"
  xmlns="http://www.w3.org/2000/svg"
  xmlns:xlink="http://www.w3.org/1999/xlink"
  xmlns:html="http://www.w3.org/1999/xhtml">
  <g id="testmeta">
    <title>Path With Round Caps</title>
    <html:link rel="author"
          title="Skef Iterum"
          href="https://github.com/skef/"/>
  </g>

  <g id="test-body-reference">
    <path style="stroke:#0A0AFF;stroke-width:2;fill:#0000FF" d="M219.021 20.2002l50 5c20.1914 2.01855 35.9795 19.0801 35.9795 39.7998c0 22.0762 -17.9238 40 -40 40c-1.35742 0 -2.69824 -0.0673828 -4.02051 -0.200195l-50 -5c-20.1914 -2.01855 -35.9795 -19.0801 -35.9795 -39.7998
c0 -22.0762 17.9238 -40 40 -40c1.35742 0 2.69824 0.0673828 4.02051 0.200195z"/>
  </g>
</svg>
```

This copies the shape from the reference test and strokes it with width 2 and a color channel difference of 10. The main advantage is that it allows one to pick the fuzzy area with the outline, as problems outside of it will have a larger channel variance. (The match on *exactly* 10 here presumably relies on anti-aliasing being turned off. Which seems to be the default? I'm a little surprised this passed in both browsers.)

Anyway if you would prefer this style I can rejigger the files.

skef commented:
@AmeliaBR I know SVG is largely a volunteer effort but it would be great if this PR could get a little attention. There's no hurry on merging it but some guidance on what shape it needs to take in order to be merged would be very helpful.

This probably appears to be backwards-looking cleanup work but it's actually a step towards a similar PR for miter-clip and arcs joins, which in turn is a step towards implementing those joins in a major browser so they can be taken off the at-risk list. Accordingly it is the join tests that are most important, and if the other tests are a distraction I can always delete them or break them out into a separate PR. 

(It is unlikely that a vendor will accept an outside contributor's implementation of something as complex as an arcs join without verification. The lack of shape-verifying tests for arcs joins is therefore part of a huge chicken-and-egg problem. That's what I'm ultimately trying to work through here.)

fsoder commented:
> @AmeliaBR I know SVG is largely a volunteer effort but it would be great if this PR could get a little attention. There's no hurry on merging it but some guidance on what shape it needs to take in order to be merged would be very helpful.

I can't speak for others, but a change of this size will come with a bit of a threshold. For the purpose of guidance it's really _way_ too large (one or two tests would probably have sufficed for that).

That being said though, I've "browsed" some parts of this commit, and it looks generally fine to me. One thing I noticed - albeit a minor detail - is that the refs use inline style for specifying fill/stroke, which seems like an unnecessary complication (refs should be as simple as possible after all). 
Personally I wouldn't have bothered with the `<g id="test-body-*">` wrapping but I know that it's used in the old testsuite and probably some new tests too - in my opinion it just adds clutter. Ditto with declaring `xlink` without using it, specifying redundant `width` and `height` on the `<svg>` root in all tests (and refs) as well the useless `id` on the same element. I'd also have saved on the vertical space (i.e not "wrap" `<meta>` elements unless they are overly long etc.). (A guiding rule would be to not add something that isn't really needed, see for example the WPT guidelines http://web-platform-tests.org/writing-tests/general-guidelines.html#be-short .)

Most of that is just nits though - except perhaps for the inline style thing, which I think is worth addressing. With that I'd be happy to approve and merge this. (Seeing that all tests pass in Firefox and Chrome - and most pass in Safari - I see no strong reason to verify every decimal point of every path. Not for basic stuff like this.)

(Note on the "alternate matching"  method: Adding additional geometry to the test doesn't sound like an improvement to me - especially if we'd still have to rely on fuzzy matching. Anti-aliasing will definitely default be "on".)

> This probably appears to be backwards-looking cleanup work but it's actually a step towards a similar PR for miter-clip and arcs joins, which in turn is a step towards implementing those joins in a major browser so they can be taken off the at-risk list. Accordingly it is the join tests that are most important, and if the other tests are a distraction I can always delete them or break them out into a separate PR.
> 
> (It is unlikely that a vendor will accept an outside contributor's implementation of something as complex as an arcs join without verification. The lack of shape-verifying tests for arcs joins is therefore part of a huge chicken-and-egg problem. That's what I'm ultimately trying to work through here.)

(I hope that "major browser" is Gecko/Firefox, because AFAIK they are the only ones to support fuzzy ref matching in their runners at the present. Given that this is mostly something that will be implemented by an underlying graphics library I suspect that more thorough testing would be required on that layer rather than on the higher levels. Also be aware that browser engines tend to have a gating process for features - even for "simple" and progressively enhancing ones like arc joins.)

skef commented:
On the nits: I was trying to follow the example of the recent Tavmjong Bah marker tests in painting/reftests. I assumed that "test-body-content" and "test-body-reference" were needed for the infrastructure but maybe that's wrong. In any case I'll unpack the inline styles and I can address the other stuff if others feel more strongly about it. 

> I can't speak for others, but a change of this size will come with a bit of a threshold. For the purpose of guidance it's really way too large (one or two tests would probably have sufficed for that).

I'm not sure whether this is a suggestion to spread the submissions over multiple PRs or to combine multiple shapes into single tests. If it is the latter, though, that's probably not a good idea for fuzzy ref-tests as the fuzziness would be cumulative and you could wind up with a bunch of close-to-exact renderings hiding one very wrong one. (Or at least that's the rationale I followed when separating them.) 

> (Note on the "alternate matching" method: Adding additional geometry to the test doesn't sound like an improvement to me - especially if we'd still have to rely on fuzzy matching. Anti-aliasing will definitely default be "on".)

I think a good case can be made for it, if it's done with care. More specifically, if you start by taking the "fuzzy contour" and pushing it alone through the infrastructure you can use its pixel count as the fuzziness count. That plus the ceiling on channel difference gives you a way of specifying a specific boundary of fuzziness, which avoids a false positive that closes matches in almost all regions while being very wrong in a particular one. (Anti-aliasing just means you nave to specify 0-10 instead of 10 for the channel difference. The fact that the example I provided passes with 10 implies to me that anti-aliasing is turned off, but maybe something else weird is going on.) 

However, that isn't my decision. I just need to know which way is preferred so I can organize the tests accordingly. 

> I hope that "major browser" is Gecko/Firefox, because AFAIK they are the only ones to support fuzzy ref matching in their runners at the present.

My current thinking is Blink/Skia because I suspect the lead time would be the shortest, and therefore the most "strategic". I've looked at Cairo but they release infrequently and I mostly see dynamic binding, so I worry it could be a year or more before anything added there is used. 

I don't understand the second part of your comment, though. As you note the tests are passing for the three major browsers (and I can assure you they're not passing "exactly").

skef commented:
Oh, and thank you @fsoder for taking the time to respond.

fsoder commented:
> On the nits: I was trying to follow the example of the recent Tavmjong Bah marker tests in painting/reftests. I assumed that "test-body-content" and "test-body-reference" were needed for the infrastructure but maybe that's wrong. In any case I'll unpack the inline styles and I can address the other stuff if others feel more strongly about it.

I did recognize the style from somewhere (and realized where). Style preferences tend to vary from person to person - myself I try to stick to short-and-minimal (per WPT guidelines). (Being "on the receiving end" of tests you don't really want to spend too much time figuring out what is relevant and what isn't.)

> > I can't speak for others, but a change of this size will come with a bit of a threshold. For the purpose of guidance it's really way too large (one or two tests would probably have sufficed for that).
> 
> I'm not sure whether this is a suggestion to spread the submissions over multiple PRs or to combine multiple shapes into single tests. If it is the latter, though, that's probably not a good idea for fuzzy ref-tests as the fuzziness would be cumulative and you could wind up with a bunch of close-to-exact renderings hiding one very wrong one. (Or at least that's the rationale I followed when separating them.)

Definitely the former (multiple commits would've worked as well as far as I'm concerned). I don't see any need to do that here though, it was just a page out of the "Reviews 101".

> > (Note on the "alternate matching" method: Adding additional geometry to the test doesn't sound like an improvement to me - especially if we'd still have to rely on fuzzy matching. Anti-aliasing will definitely default be "on".)
> 
> I think a good case can be made for it, if it's done with care. More specifically, if you start by taking the "fuzzy contour" and pushing it alone through the infrastructure you can use its pixel count as the fuzziness count. That plus the ceiling on channel difference gives you a way of specifying a specific boundary of fuzziness, which avoids a false positive that closes matches in almost all regions while being very wrong in a particular one. (Anti-aliasing just means you nave to specify 0-10 instead of 10 for the channel difference. The fact that the example I provided passes with 10 implies to me that anti-aliasing is turned off, but maybe something else weird is going on.)
> 
> However, that isn't my decision. I just need to know which way is preferred so I can organize the tests accordingly.

I'm not too familiar with the fuzzy matching, but AFAIU "10" and "0-10" would be the same. That might explain why that passes. But, yeah, my preference would be to have "less" (especially if fuzzy matching is still a requirement).

> > I hope that "major browser" is Gecko/Firefox, because AFAIK they are the only ones to support fuzzy ref matching in their runners at the present.
> 
> My current thinking is Blink/Skia because I suspect the lead time would be the shortest, and therefore the most "strategic". I've looked at Cairo but they release infrequently and I mostly see dynamic binding, so I worry it could be a year or more before anything added there is used.

And the process the most arduous =). On the "web engine" side this is just a (fairly trivial) plumbing exercise though. (FWIW, I'm not sure if Gecko uses Cairo anymore - ISTR they use Skia as well when they don't use something else.)

> I don't understand the second part of your comment, though. As you note the tests are passing for the three major browsers (and I can assure you they're not passing "exactly").

AFAIK, Gecko is the only engine that uses the WPT runner (or something close to it) for their CI-testing. Blink does not, and that CI-test-runner does not support fuzzy matching. So what's on wpt.fyi (or achieved by running the `wpt` tool) does not necessarily reflect what browsers actually "see" (i.e it's not the source of truth as such). So, while having WPT tests will be a plus, it won't really help much because they'll likely just need to be skipped/disabled.

skef commented:
> I did recognize the style from somewhere (and realized where). Style preferences tend to vary from person to person - myself I try to stick to short-and-minimal (per WPT guidelines). (Being "on the receiving end" of tests you don't really want to spend too much time figuring out what is relevant and what isn't.)

Ah, well, I just found the geometry circle tests which are more minimal, so while I'm editing the files I'm following most of your suggestions (but I'm leaving the body/reference stuff). 

> So, while having WPT tests will be a plus, it won't really help much because they'll likely just need to be skipped/disabled.

I doubt that will ultimately matter. My purpose in this PR is to establish a pattern for writing miter-clip and arcs join tests. The solution to the chicken-and-egg problem is the eventual acceptance, presumably after some negotiation and changes, of those tests. That establishes an objective standard I can point to, even if it's just pointing at my own run of `wpt` or attaching the files for visual comparison. That's light-years ahead of linking to the spec and plaintively insisting that I've followed it. There's no particularly tricky math in arcs but fitting it all together turns out to be a pain and there are lots of ugly details and exceptions.

skef commented:
>  (FWIW, I'm not sure if Gecko uses Cairo anymore - ISTR they use Skia as well when they don't use something else.)

I think *something* still does but I don't remember what -- I was mostly looking at it because Inkscape uses it. I've seen claims that Inkscape can qualify as the "other implementation" already but that seems questionable -- the geometric support is there (sort of) but it's not actually hooked up to the SVG parser as far as I can tell. 

However, that's not important for the "at risk" problem because if it doesn't qualify then FontForge will.

skef committed:
Simplify/remove unneeded or unused tags

skef commented:
Alright, I simplified the files. Let me know if I should also delete the corresponding manual tests as part of the PR.

fsoder commented:
> I doubt that will ultimately matter. My purpose in this PR is to establish a pattern for writing miter-clip and arcs join tests. The solution to the chicken-and-egg problem is the eventual acceptance, presumably after some negotiation and changes, of those tests. That establishes an objective standard I can point to, even if it's just pointing at my own run of `wpt` or attaching the files for visual comparison. That's light-years ahead of linking to the spec and plaintively insisting that I've followed it. There's no particularly tricky math in arcs but fitting it all together turns out to be a pain and there are lots of ugly details and exceptions.

Not wanting to extend this tangent any further but... as a long-time Blink contributor I think I can speak with some authority on the subject. "Pointing at" external tests (or their results) is not how that story goes, but as someone who could well be performing the review there as well, basic regressions tests + good tests in Skia would likely be sufficient. But this is starting to become too irrelevant for this PR, so let's end it at that (and I wish you good luck with that).

fsoder commented:
> Alright, I simplified the files. Let me know if I should also delete the corresponding manual tests as part of the PR.

I think that could be handled separately (just referencing this PR).

fsoder reviewed (approved):
LGTM. Let's allow some additional time for others to comment. Ping the PR again after that and we can merge it.

skef commented:
@AmeliaBR If I'm reading the logs correctly the alternate fuzzy matching convention was something [you favored](https://www.w3.org/2019/09/18-svg-irc#T02-32-50) or at least noted. This would be a good time and context to weigh in on it again. No need to re-review to express a preference on that.

AmeliaBR commented:
Thanks for the PR & to @fsoder for the detailed review.

I second all of Frederick's comments about style & simplification (and smaller PRs in future, at least until there's clear agreement on a template), but I think this is definitely an improvement that is worth merging now.

That said: in the long term, we do want to avoid fuzzy testing whenever possible. In part, this is because of the issues with the browser's internal test suites that Frederick mentioned, in part it is because of the concern mentioned in the OP about fuzzy tests sometimes fuzzing the wrong things.

If you want to update the PR now, that'd be great, but we could also merge it as-is and then update later.

> discussion on one of the mailing lists of an alternate matching heuristic

That was my proposal & I guess it wasn't transcribed clearly. My intention was that the extra stroked shape would be on a separate element that is added to _both_ the test and reference files. It would be a contrasting color that covers up the areas where fuzzy aliasing might occur.

Here's some example markup for a highly simplified variation of one of your tests:

```xml
<!-- test -->
    <path stroke="blue" stroke-width="80"
          d="M 215 60 h 30"/>
    <path stroke="black" stroke-width="2" fill="none"
          d="M215 20 L 245 20 L 245 100 L215 100 Z"/>

<!-- ref -->
    <path fill="blue"
          d="M215 20 L 245 20 L 245 100 L215 100 Z"/>
    <path stroke="black" stroke-width="2" fill="none"
          d="M215 20 L 245 20 L 245 100 L215 100 Z"/>
 ```

Note that the second element (the outline) is exactly identical in both, and covers up one pixel inside and one pixel outside the shape that is actually being tested. It shouldn't require any fuzziness in the matching, because the black outline should always be in the exact same position, and composited over solid blue on one side and empty background on the other.  If we end up with some tests (e.g., for more extreme curved shapes) where a 2px-wide fuzz zone isn't enough to ensure exact references, we could use a wider stroke width as needed.

AmeliaBR commented:
>> Let me know if I should also delete the corresponding manual tests as part of the PR.
>
> I think that could be handled separately (just referencing this PR).

Either works. The goal is to eventually empty out the import folder by updating those tests to WPT format, so deleting the files is important to help us recognize that they've already been replaced!

skef commented:
@AmeliaBR Thanks for the response. Note that the tone of my email was intended to be gently chiding -- if it came across stronger sorry about that. 

Let me try to make the case for the alternate style I posted above first, and if you don't agree I'll restructure the tests in the way you describe. 

Your proposal is to draw identical lines of a contrasting color *over* the aliased border and match without fuzziness. My suggestion is to draw a line of a similar but not exact color *under* the aliased border and fuzzy match on the max difference between the color channels and the number of pixels covered by the "under-line" when drawn alone. 

Because the second system limits the color channel difference both of these conventions control the "area of fuzziness". The main advantage of the first proposal is that it avoids having to specify fuzziness parameters at all. I see the main advantage of the second proposal as supplying a passing cutoff while still giving the tester some measure of how close they matched, in the form of the pixel difference count. I think that's useful information. 

Let me know about that and whether you want a stroke width of 2 or 1.2 for the shape (given the enlargement factor over the viewbox size for these tests) and I'll update them.

skef commented:
Oh, also: I would be happy if these tests could wind up being a model for future tests, and you're the authority, so if there is additional cleanup you would like to see beyond my first pass just let me know the specifics.

skef commented:
Hmm -- on second thought the understroke system is at least difficult and perhaps impossible to layer into shapes that are both stroked and filled. I'll work up a draft with 1.2 width overlines as suggested.

AmeliaBR commented:
> a stroke width of 2 or 1.2 for the shape (given the enlargement factor over the viewbox size for these tests)

Oh, good catch — I wasn't thinking about that. Given that we're not explicitly testing viewBox scaling in any of these, it might be worth removing the viewBox attribute so they render at the raw scale. Or replace it with width/height attributes (I wonder if the lack of those is what's ruining GitHub's preview function, or if they're still having server issues…)

> I see the main advantage of the second proposal as supplying a passing cutoff while still giving the tester some measure of how close they matched, in the form of the pixel difference count.

That is indeed an interesting argument, but I think it is cancelled out by the fact that Chromium (and WebKit?) doesn't have fuzzy matching support on their internal test runner, and so these tests would be excluded from their automatic regression tests.

> if there is additional cleanup you would like to see beyond my first pass

I would drop the `<g id="testmeta"> wrapper around the metadata. A `<g>` is for rendered content. A `<metadata>` element would make more sense here, but isn't required. (Some existing tests use it, some go for the minimal approach.)

But in particular, the SVG `<title>` should be the first direct child of the `<svg>`, not inside a wrapper.

Titles should explain what makes this test different from others. E.g., what is the difference between “Circle 2” test versus “Circle 4”.  If it needs more explanation than a short title can hold, add a [`<meta name="assert">](https://web-platform-tests.org/writing-tests/css-metadata.html#test-assertions) following the CSS conventions.

If you can't come up with a good explanation about what makes a test unique, maybe we don't need one of those tests. (We don't want redundant tests, as it slows down all the tool chains that test browser builds.)

As I'm looking at them now, I think we probably don't need quite this many basic shapes tests.  For a basic test of geometry (comparing basic shapes to their path equivalents), we probably want something very small, something large, something filled, something stroked, but beyond that each test needs to be testing a unique feature.

skef committed:
Switch to overline non-fuzzy style (except miterlimit)
Further simplifcation

skef commented:
I've now updated all the tests to use the overline convention except for the miterlimit test (which faces a reflex problem). I used a width of 1.5 because 1.2 seemed a little too aggressive. I've also simplified as suggested. 

Not all of these pass but the particular failures aren't common across browsers and suggest they might be something to examine. 

The duplicate shape tests are just copied from the imported files. Tell me which to delete and/or change and I'll do so.

skef commented:
> Given that we're not explicitly testing viewBox scaling in any of these, it might be worth removing the viewBox attribute so they render at the raw scale. Or replace it with width/height attributes (I wonder if the lack of those is what's ruining GitHub's preview function, or if they're still having server issues…)

You're the expert ([e.g.](https://css-tricks.com/scale-svg/)) but the idea of matching differing geometries without precise placement in a viewBox makes me nervous.

skef committed:
Remove miterlimit tests

skef commented:
Actually, given that the next step for me is to write some miter-clip tests I just deleted the current miterlimit test. I'll try to come up with a better convention for these and we can consider it in the next PR.

skef committed:
Delete equivalent "import" manual tests

AmeliaBR commented:
>  the idea of matching differing geometries without precise placement in a viewBox makes me nervous

So long as both files are the same (e.g., no viewBox), then the reference system is the same: pixels measured from the top-left.  And those pixels should match up with device pixels in the snapshots, which should hopefully help minimize some of the aliasing issues.

> the particular failures aren't common across browsers and suggest they might be something to examine

Checking the visual diffs, most of the problems are in circles & ellipses. Since the spec gives browsers some flexibility in how those are converted to Bézier curves, it might be OK to add a slightly larger “fuzz” allowance (AKA, use a wider masking stroke) for arcs, circles, and ellipses.

The other fails are:

- strokefill-003.svg, in Chromium and Firefox the mismatched pixels are at the same corners, and the values differ in whether they are pure gray or slightly greenish, so the color is bleeding through the black outline.

- strokefill-001.svg in WebKit, an entire diagonal edge is compositing slightly differently (very slightly: the colours are `#010101` vs `#020202`, with no bias towards the color on the lower layer). 

These differences are probably partly based on when the browser rounds off colors from solidly-colored sub-pixels to partially transparent full pixels, as it composites different layers together.

To minimize arbitrary differences, I'd recommend changing [the ref test](https://github.com/web-platform-tests/wpt/pull/22791/files#diff-d3ea7bcd7230fe26c7d3a0a1cec3abe3) to use the exact same (separate) elements for the black stroke as in the test file, instead of trying to combine those strokes with the shapes that are providing the fill color (or the stroke color as fill color).

skef commented:
@AmeliaBR I'm going to update one or two tests as I understand your latest feedback and let you verify or correct me. Then I'll do the rest.

skef committed:
Change viewBox to width/height
Switch overline strokewidth to 2
Separate stroke and fill in reference svg

skef commented:
OK, take a look at `strokefill-003` and `circle-003`. 

I deleted `circle-002` because the stroked border was effectively too narrow for the test system -- it was more or overwritten by the overline. 

Concerning "Since the spec gives browsers some flexibility in how those are converted to Bézier curves, it might be OK to add a slightly larger “fuzz” allowance (AKA, use a wider masking stroke) for arcs, circles, and ellipses.": Based on my work on another tool the slop in the stroke width is already pretty high for that sort of consideration. Vendors will be happy with passing tests but you don't want to set the margins so large as to hide things they might want to look at.

skef commented:
I'm not familiar with your infrastructure but from the error page and logs it doesn't look related to the PR content.

AmeliaBR commented:
Yes, those files match my suggestions.

Not sure how to get those checks to re-run, but they'll do so automatically if you push another commit.

skef committed:
Adapt to width=480/height=320 and overline stroke-width=2.4

skef commented:
All files updated. With width=480 I had to switch to stroke-width=2.4 on the overline to get almost all tests passing. Still a few low-pixel-count failures.

fsoder commented on svg/path/reftests/strokefill-001-ref.svg:
```diff
@@ -0,0 +1,13 @@
+  <path stroke="none" fill="green" d="M87.1719 77.1719l-60 60l-1.17188 1.17285v1.65527v60v1.65527l1.17188 1.17285l60 60l1.17285 1.17188h1.65527h60h1.65527l1.17285 -1.17188l60 -60l1.17188 -1.17285v-1.65527v-60v-1.65527l-1.17188 -1.17285l-60 -60
+  l-1.17285 -1.17188h-1.65527h-60h-1.65527z"/>
+  <path stroke="none" fill="yellow" d="M34 141.655l57.6553 -57.6553h56.6895l57.6553 57.6553v56.6895l-57.6553 57.6553h-56.6895l-57.6553 -57.6553v-56.6895z"/>
+  <path stroke="black" stroke-width="2.4" fill="green" d="M87.1719 77.1719l-60 60l-1.17188 1.17285v1.65527v60v1.65527l1.17188 1.17285l60 60l1.17285 1.17188h1.65527h60h1.65527l1.17285 -1.17188l60 -60l1.17188 -1.17285v-1.65527v-60v-1.65527l-1.17188 -1.17285l-60 -60
+  l-1.17285 -1.17188h-1.65527h-60h-1.65527z"/>
+  <path stroke="black" stroke-width="2.4" fill="yellow" d="M34 141.655l57.6553 -57.6553h56.6895l57.6553 57.6553v56.6895l-57.6553 57.6553h-56.6895l-57.6553 -57.6553v-56.6895z"/>
```
Shouldn't these two have fill="none" and be in the opposite order to match the test?

skef replied:
The order of the overlines should be irrelevant -- even if they overlap the result should be "cumulative". However, the fills do need to be removed, which I've just done.

AmeliaBR replied:
They should still match, since the whole point is that they are supposed to be identical & we're trying to minimize any chance of arbitrary compositing differences.

But as it turns out, this isn't one of the ones currently triggering a difference.

skef replied:
Alright, I matched them up.

fsoder commented:
Are we ready enough to merge this? (Modulo the above nit.)

skef committed:
Remove spurious fills

AmeliaBR commented:
> Are we ready enough to merge this?

We're still getting some “false” fails from anti-aliasing differences. Not sure if that should be a blocker for merging.

skef commented:
@AmeliaBR The current error margin is generous and while I can't swear to it I have high confidence in the geometry matching. The fact that most of the failures are browser-specific supports that. 

Rather than just boosting the margin further now, might it be better to do that (or not) in discussion with the groups in question, if and when they take a look? 

(Another possibility is to reintroduce a minimal amount of fuzziness -- something like 20 pixels -- rather than increasing the whole error margin.)

skef committed:
Match order of overline in test and reference

AmeliaBR commented:
> Rather than just boosting the margin further now

Yes, I agree. you're already giving more than a pixel of margin on either side of the edges, so increasing that is starting to cancel out the purpose of the test. But I'm not sure what else to do, other than adding a very small fuzzy allowance back in (e.g.,±2 per colour channel, on <20 pixels).

We could maybe merge & see if the failures get more people looking at the tests & offering suggestions…

@fsoder, what do you think?

skef commented:
@AmeliaBR and @fsoder what's the status on this?
